extensions/caffeine/deployment/src/main/java/io/quarkus/caffeine/deployment/CaffeineProcessor.java

```diff
@@ -0,0 +1,20 @@
+package io.quarkus.caffeine.deployment;
+
+import java.io.IOException;
+
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.substrate.ReflectiveClassBuildItem;
+
+public class CaffeineProcessor {
+
+    @BuildStep
+    ReflectiveClassBuildItem cacheClasses() throws IOException {
+        //todo: how to we want to handle this? There are a lot of different cache classes
```

Comment:
Please let me know if there are any design hints to assist here. The codegen is intended to reduce memory overhead at the cost of jar bloat, with dynamic class loading to so that most are unused. That makes a lot of sense on HotSpot, but isn't friendly to AOT compilation. There are tons of configuration options where caches might be large, so it's a good optimization. Unfortunately I don't know of a better solution that is also Graal friendly...

Reply:
Is there some way to figure out what classes are needed based on config? We could just index them all but that is not a great solution, as it results in unnecessary stuff in the native image.

Another possibility might be to make sure that the cache is created in the static init phase. As this runs in a normal JVM we might be able to just avoid the problem entirely.

Reply:
For static phase `Micronaut` recently switched to embedding my older [ConcurrentLinkedHashMap](https://github.com/ben-manes/concurrentlinkedhashmap) library. That's Java 6 based and only focused on concurrent LRU.

The classes are generated during compilation by JavaPoet. The `Cache` class name is determined by configuration in `LocalCacheFactory#newBoundedLocalCache` and the `Node` class names (entry) in `NodeFactory#newFactory`. Both factory classes are generated, so you would best look in your IDE. The config is translated into short names and an instance reflectively loaded.